thunder/executors/torchex.py

```diff
@@ -2190,6 +2182,9 @@ def is_float_type(self, input):
 
 
 def _copy__impl(copy_from, copy_to):
+    cd = get_compile_data()
+    if cd is not None and cd.is_grad_enabled and copy_to.is_leaf and copy_to.requires_grad:
+        raise RuntimeError("a leaf Variable that requires grad is being used in an in-place operation.")
```

Comment:
I am wondering if Symbol `copy_` in `thunder/torch/__init__.py` is more appropriate location for the check. 

https://github.com/Lightning-AI/lightning-thunder/blob/60f3ee1ec536ee8d6fdef503af54525e0a3978a4/thunder/torch/__init__.py#L1961-L1963

Reply:
`a` and `b` are proxies and it it not clear to me if a proxy knows that it is a leaf.

Reply:
They do not. It's only a PyTorch concept that's available at runtime inside `_copy__impl`.

Reply:
Indeed, compile_data was not available, but now it should be with the added context manager in thunder/__init__.py

Reply:
I think this is still incorrect because as discussed in https://github.com/Lightning-AI/lightning-thunder/issues/1486, the value of compile_data.is_grad_enabled here would be that of last updated state which can lead to incorrectness when used outside of tracing context.

We can see the discrepancy here.
```python
import torch
import thunder

x = torch.randn(3, 3, requires_grad=True)

@torch.no_grad
def fn(x):
  return x.add_(1)

fn(x)  # This works

thunder.jit(fn)(x)  # This raises error
```

So, whether the `copy` is in no_grad region needs to be captured during the tracing time.

Reply:
Right, this is why I created the other issue.  This PR fixes the leaf/grad issue when there is no annotation.  When there is an annotation, another approach is required.  This other approach may or may not involve using compile data in `_copy__impl`.

As far as I understand, compile data is the medium for passing around data such as whether grad is enabled.  But as the other issue points out, compile data reflects the end state of a function call and not the "live" state, at least at the time it reaches `_copy__impl`.  So I'm left with the questions "are there other mechanisms for passing around whether grad is enabled?" "where else in the execution is it simultaneously knowable that a (1) leaf tensor (2) requiring grad is being (3) copied when (4) grad is enabled?" "is it feasible/desirable to make the compile data more dynamic?"  "is there a way to context-manage the tensors so that their `requires_grad` flags are set to `False` when the interpreter sees `torch._C._set_grad_enabled(False)`, and then later restored, thereby obviating the need for the compile data for this check?"  Do you have suggestions for a fix that addresses both issues? Or can we close out this issue and move the discussion to the more involved issue?

Reply:
So to tackle - leaf tensor requiring grad being copied into when grad is enabled, I think similar to a previous commit, 
we can update `prims.copy` to take a argument `is_grad_enabled`. With this, `ltorch.copy` will query `cd.is_grad_enabled` and call `prims.copy` by also passing this argument.

https://github.com/Lightning-AI/lightning-thunder/blob/9de543417260bd32740e2d7bd0763a5871fdee5c/thunder/torch/__init__.py#L1984-L1986

With these changes, the `copy_impl`'s signature will also change to accept `is_grad_enabled` and it will be called at runtime with a tensor which we can query if it is a leaf and also whether grad was enabled or not when calling that particular copy. Wdyt @beverlylytle?

Though, I am curious if there is another approach to this - cc: @IvanYashchuk

Reply:
Let's see what the CI thinks.

Reply:
I agree with modifying `thunder.torch.copy` to query `cd.is_grad_enabled` and passing that to `prims.copy`.